[fuchsia_remote_debug_protocol] Adds the fuchsia remote debug protocol library.

awdavies opened the pull request:
Includes very basic functionality to connect to a VM instance and
determine the active Flutter Views running on the machine.

awdavies committed:
[flutter_fuchsia] Adds flutter_fuchsia.

Includes very basic functionality to connect to a VM instance and
determine the active Flutter Views running on the machine.

awdavies committed:
[flutter_fuchsia] Adds flutter_fuchsia.

Includes very basic functionality to connect to a VM instance and
determine the active Flutter Views running on the machine.

yjbanov commented on packages/flutter_fuchsia/lib/example_main.dart:
```diff
@@ -0,0 +1,21 @@
+import 'dart:io';
```
This file should probably go under `bin/` next to the `lib/` directory (https://www.dartlang.org/tools/pub/package-layout). We don't want anything from here to accidentally show up in the API docs.

awdavies replied:
Done.

yjbanov commented on packages/flutter_fuchsia/lib/example_main.dart:
```diff
@@ -0,0 +1,21 @@
+import 'dart:io';
```
Missing license header (see existing Dart files in this repo)

awdavies replied:
Done.

yjbanov commented on packages/flutter_fuchsia/lib/example_main.dart:
```diff
@@ -0,0 +1,21 @@
+
+import 'flutter_fuchsia.dart';
+
+Future<Null> main(List<String> args) async {
```
Let's add some dartdocs clarifying why this example is here and what it is demonstrating.

awdavies replied:
Done.

yjbanov commented on packages/flutter_fuchsia/lib/example_main.dart:
```diff
@@ -0,0 +1,21 @@
+  print(views);
+
+  // Program hangs here, so force an exit.
+  exit(0);
```
This probably indicates a flaw in the API. You can use observatory to find any dangling "ports" (web-sockets maybe?). Perhaps the `FlutterView` class needs a `close` method to clean things up.

awdavies replied:
Yeah, I'm a bit stumped on what causes this.  In the flutter_tools implementation they run through some shutdown hooks, then run a null completer after some timer expires, which calls the exit function explicitly.

I'm not sure if they have a similar problem, but I'll take a look at observatory.

awdavies replied:
Fixed!

yjbanov commented on packages/flutter_fuchsia/lib/flutter_fuchsia.dart:
```diff
@@ -0,0 +1,12 @@
+// Copyright 2018 The Fuchsia Authors. All rights reserved.
```
I think we want to use Flutter's license headers for all packages in this repo. /cc @Hixie

yjbanov commented on packages/flutter_fuchsia/lib/flutter_fuchsia.dart:
```diff
@@ -0,0 +1,12 @@
+/// Provides an API to test Flutter applications on Fuchsia devices and
+/// emulators.
+///
+/// The application runs in a separate process from the actual test.
```
Technically it can run in the same process, so long as it can connect to the application via a web-socket. But we should probably still say that _usually_ it runs in a separate process or even on a separate computer.

awdavies replied:
Updated.

yjbanov commented on packages/flutter_fuchsia/lib/src/dart/fuchsia_dart_vm.dart:
```diff
@@ -0,0 +1,99 @@
+  /// connect.
+  static Future<FuchsiaDartVm> connect(Uri uri) async {
+    final Stopwatch timer = new Stopwatch()..start();
+    if (uri.scheme == 'http') uri = uri.replace(scheme: 'ws', path: '/ws');
```
The contents of the `if` block should go on the next line (or better, use the `{ }` block). Also check out our addendum to the Dart style guide: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo.

awdavies replied:
I think I'll just put it in brackets.  All of this code autoformats itself with dart-fmt when I write.

awdavies replied:
Done.

yjbanov commented on packages/flutter_fuchsia/lib/src/dart/fuchsia_dart_vm.dart:
```diff
@@ -0,0 +1,99 @@
+        await new Future<Null>.delayed(_kReconnectAttemptInterval);
+        return FuchsiaDartVm._attemptConnection(uri, runningTimer);
+      } else {
+        log.critical('Connection to Fuchsia\'s Dart VM timed out.');
```
Let's print the URL that it tried to connect to.

awdavies replied:
Done.

yjbanov commented on packages/flutter_fuchsia/lib/src/dart/fuchsia_dart_vm.dart:
```diff
@@ -0,0 +1,99 @@
+    }
+  }
+
+  /// Invokes a raw RPC command with the VM service.
```
nit: "raw JSON RPC" to be more specific.

awdavies replied:
Done.

yjbanov commented on packages/flutter_fuchsia/lib/src/dart/fuchsia_dart_vm.dart:
```diff
@@ -0,0 +1,99 @@
+    return _peer.sendRequest(function, params ?? {});
+  }
+
+  /// Returns a list of flutter views by name. If the name of a flutter view
```
Make the first sentence a standalone paragraph (https://www.dartlang.org/guides/language/effective-dart/documentation#do-make-the-first-sentence-a-standalone-paragraph)

awdavies replied:
Done.

yjbanov commented on packages/flutter_fuchsia/lib/src/dart/fuchsia_dart_vm.dart:
```diff
@@ -0,0 +1,99 @@
+  /// Returns a list of flutter views by name. If the name of a flutter view
+  /// cannot be determined (there is no associated isolate), then the flutter
+  /// view's ID will be added instead.
+  Future<List<String>> listFlutterViewsByName() async {
```
nit: the name of this method makes it sound like it sorts or groups by name; perhaps `listFlutterViewNames`?

awdavies replied:
Done.

yjbanov commented on packages/flutter_fuchsia/lib/src/dart/fuchsia_dart_vm.dart:
```diff
@@ -0,0 +1,99 @@
+
+  /// Returns a list of flutter views by name. If the name of a flutter view
+  /// cannot be determined (there is no associated isolate), then the flutter
+  /// view's ID will be added instead.
```
Somewhere in the API docs we should clearly define what is a "flutter view", "ID" and other Fuchsia-specific concepts. It's ok if there's no clear (or stable) definition yet, but anything that clarifies it for the user should be useful.

awdavies replied:
To be totally frank, I don't think I know enough about the specifics of those things yet to clearly define them (mainly the ID.  It is something I'm currently using because it is available and a string).

yjbanov commented on packages/flutter_fuchsia/lib/src/dart/fuchsia_dart_vm.dart:
```diff
@@ -0,0 +1,99 @@
+    final List<Map<String, dynamic>> flutterViews = rpcResponse['views'];
+    for (Map<String, dynamic> flutterView in flutterViews) {
+      Map<String, dynamic> isolate = flutterView['isolate'];
+      String id = flutterView['id'];
```
Local variables that do not change after assignment should be marked as `final`. In fact, the analyzer should have flagged this. If not, we should check that `flutter analyze --flutter-repo` captures this new package along with the others.

awdavies replied:
Done.  Also I ran `flutter analyze` with/without the dashes and found no information about this package.

yjbanov commented on packages/flutter_fuchsia/lib/src/dart/fuchsia_dart_vm.dart:
```diff
@@ -0,0 +1,99 @@
+      } else if (id != null) {
+        viewNames.add(id);
+      } else {
+        _log.severe('Unable to find view name for the following JSON structure '
```
`severe` usually means that the process will halt, but here it attempts to keep on trucking. Consider using a less severe log level, such as warning or error.

awdavies replied:
Hmmm, I guess I'm unsure of the semantics for the logger.  It doesn't have an error function in the pub docs (which is why I switched it to severe).  I'll move things over/change behavior for the actually-severe logs, though.

awdavies replied:
Fixed.

yjbanov commented on packages/flutter_fuchsia/lib/src/dart/fuchsia_dart_vm.dart:
```diff
@@ -0,0 +1,99 @@
+        if (name != null) {
+          viewNames.add(name);
+        } else {
+          _log.severe('Unable to find name for isolate "$isolate"');
```
The docs say that if the name cannot be determines, the ID will be used, but the implementation seems to contradict it.

Also, nit: here and elsewhere in this PR, end log messages with a period.

awdavies replied:
Ah, I think I forgot to update the documentation.  If there is no associated isolate, it will contain an ID.

awdavies replied:
Fixed.

yjbanov commented on packages/flutter_fuchsia/lib/src/flutter_views.dart:
```diff
@@ -0,0 +1,187 @@
+import 'dart/fuchsia_dart_vm.dart';
+import 'fuchsia_device_command_runner.dart';
+
+final String ipv4Loopback = InternetAddress.LOOPBACK_IP_V4.address;
```
Can this be private?

awdavies replied:
Done.

yjbanov commented on packages/flutter_fuchsia/lib/src/flutter_views.dart:
```diff
@@ -0,0 +1,187 @@
+
+final String ipv4Loopback = InternetAddress.LOOPBACK_IP_V4.address;
+
+final Logger _log = new Logger('flutter_fuchsia::flutter_views');
```
This logger uses a different name scheme from the one above. Let's pick one.

awdavies replied:
Done.

yjbanov commented on packages/flutter_fuchsia/lib/src/flutter_views.dart:
```diff
@@ -0,0 +1,187 @@
+final ProcessManager _processManager = new LocalProcessManager();
+
+/// Persistent VM service cache to avoid repeating handshakes across function
+/// calls.
```
Let's document what the key in this map is.

awdavies replied:
Done.

yjbanov commented on packages/flutter_fuchsia/lib/src/flutter_views.dart:
```diff
@@ -0,0 +1,187 @@
+final HashMap<int, FuchsiaDartVm> _fuchsiaDartVmCache =
+    new HashMap<int, FuchsiaDartVm>();
+
+/// Returns a list of flutter views for the given [ipv4Address].
```
nit: Flutter style guide (deliberately in contradiction to the upstream dart style) uses backticks to refer to parameters rather than square brackets.

awdavies replied:
Done.

yjbanov commented on packages/flutter_fuchsia/lib/src/flutter_views.dart:
```diff
@@ -0,0 +1,187 @@
+
+/// Returns a list of flutter views for the given [ipv4Address].
+/// TODO(awdavies): just returns flutter view names. Needs to return
+/// FlutterView objects of some kind that contain a JSON RPC peer.
```
To reduce future breaking changes, let's start with a `FuchsiaFlutterView` class that has a single field `String name`. In the future we can add more methods to it without breaking existing clients.

awdavies replied:
sgtm

awdavies replied:
Done.

yjbanov commented on packages/flutter_fuchsia/lib/src/flutter_views.dart:
```diff
@@ -0,0 +1,187 @@
+  final List<String> lsOutput = await runner.run('ls /tmp/dart.services');
+  final List<int> ports = <int>[];
+  for (String s in lsOutput) {
+    final String trimmed = s.trim();
```
It would be useful to provide an example string this code is trying to parse in a comment.

awdavies replied:
good idea

awdavies replied:
Done.

yjbanov commented on packages/flutter_fuchsia/lib/src/flutter_views.dart:
```diff
@@ -0,0 +1,187 @@
+
+// Instances of this class represent a running ssh tunnel from the host to a
+// VM service running on a Fuchsia device. [process] is the ssh process running
+// the tunnel and [port] is the local port.
```
Use `///` for dartdocs.

awdavies replied:
Done.

yjbanov commented on packages/flutter_tools/lib/src/vmservice.dart:
```diff
@@ -78,7 +78,8 @@ class VMService {
         final bool pause = params.asMap['pause'] ?? false;
 
         if (isolateId is! String || isolateId.isEmpty)
-          throw new rpc.RpcException.invalidParams('Invalid \'isolateId\': $isolateId');
```
This file seems to only contain formatting changes. Can we revert this file?

awdavies replied:
Yes.  Not sure how that happened :confused:

awdavies replied:
Done.

yjbanov commented on packages/flutter_fuchsia/pubspec.yaml:
```diff
@@ -0,0 +1,11 @@
+name: flutter_fuchsia
```
You will need to run `flutter update-packages --force-upgrade`, which will populate transitive dependencies of this package and make sure the dependencies' version constraints are consistent with the rest of the repo.

awdavies replied:
I believe this is fixed now.  We'll see if things run correctly in Travis, I guess.

yjbanov commented:
The high-level structure of this package LGTM. Let's do a round of design review, then style review. Let's also make sure the package is configured consistently with the other packages, e.g. transitive dependencies in `pubspec.yaml`.

Also, let's add tests! 😃

awdavies committed:
Added transitive deps

awdavies committed:
Addressing most example_main.dart comments.

Still need to address the issue of hanging when the program should exit.

awdavies committed:
Addresses comments for fuchsia_dart_vm.dart

Also adds in a couple fixes to logged messages in flutter_views.dart

awdavies committed:
Reverts flutter_tools code and addresses comments.

Addresses comments for flutter_views.dart

awdavies committed:
Added in missing private variable refactor.

awdavies committed:
Updating library comments.

awdavies committed:
Adds a fuchsia flutter view class.  Updates logging code.

awdavies committed:
Formatting.

awdavies commented:
Most code is addressed.  Still figuring out what makes exiting the program hang (having some issues with observatory).

Hixie commented:
I'm concerned about this being a top-level package. Can you elaborate on the use case and why this is a package? Do we expect people to import it?

awdavies commented:
@Hixie the package is intended to function similar to the flutter_driver, but from the context of running in a Fuchsia environment.  This would require significant refactoring of the existing flutter_driver, due to the fact that there are often multiple instances of the Dart VM running at a given time, and that an application can be composed of multiple isolates across multiple Dart VM's.

We would expect people to import this, yes.  This change is mostly a dummy one for myself to figure out the basics of connecting/talking to the Fuchsia Dart VM (and to work out any potential issues I might run into with things like flutter analyze, continuous integration, etc) before trying to tackle a robust API.

Hixie commented:
Ah, I see. Maybe it should be called something like `multiinstance_driver`? I'm worried that `flutter_fuchsia` really doesn't convey what it is; the `driver` part of `flutter_driver` is really the important part, not the `flutter` part.

awdavies commented:
sgtm.  @yjbanov thoughts?

Hixie commented:
Also unless this is very fuchsia-specific I recommend keeping the API agnostic about the OS, because that way if we ever have a similar setup on another platform we can re-use this without the awkward problem of the classes referring to another OS. :-)

awdavies committed:
Fixes hanging issues.

awdavies committed:
Addresses `flutter analyze`

awdavies commented:
@Hixie To double check again with a comment Yegor made: should the License header be written under the Flutter team, or under the Fuchsia team?  I figure given the OS agnostic comment it'd be Copyrighted under the Flutter team?

@yjbanov PTAL at the code (also check up on Ian's comments about the purpose of this package).

yjbanov commented:
@Hixie, all of the OS-agnostic bits will indeed go in `package:flutter_driver`. This new package would only contain Fuchsia-specific parts (How to talk to a Fuchsia device over USB and SSH. How to track the lifecycle of processes/stories/mods running in Fuchsia. How to enable/disable the driver extension inside a Flutter app running on Fuchsia.). One way to look at it is this is "adb for Fuchsia". It can indeed be imported and used independently of `flutter_driver` and `flutter_tools` (perhaps the Fuchsia infra team can use it for some testing). I still think that this package should live over in the Fuchsia source (similar to how adb is developed as part of the Android SDK), and managed and tested using Fuchsia infra (`jiri` and so on). But I failed to convince the Fuchsia team to do this. Developing it as part of Flutter but as a standalone package seemed like a good compromise.

yjbanov commented:
- [ ] let's add tests
- [ ] when you ran `flutter update-packages --force-upgrade`, did it update other pubspecs? It usually does

awdavies commented:
- Will do!
- Yes, it did. I wasn't sure whether to include changes to other pubspecs as part of this change.

awdavies commented:
/CC @krisgiesing here's the start to the driver impl we discussed

yjbanov commented:
> I wasn't sure whether to include changes to other pubspecs as part of this change.

Yes, they should be included. The analyzer will complain if the pubspecs have different version numbers.

awdavies committed:
Updates VM connection impl (makes it easier to test).

Changes underyling FlutterFuchsia implementation to use a Driver class
similar to Flutter's existing driver library.

Hixie commented:
I don't really understand what this package is. I can't work out if it's flutter-specific, fuchsia-specific, driver-specific, or what.

Can we meet before this lands to discuss it?

Hixie commented:
> @Hixie To double check again with a comment Yegor made: should the License header be written under the Flutter team, or under the Fuchsia team? I figure given the OS agnostic comment it'd be Copyrighted under the Flutter team?

All code in this repo should have the exact same copyright license. Just copy it from another file, do not change anything except (optionally) the year (if and only if you created the entire file from scratch).

Hixie commented:
Ok so @yjbanov explained some of the background to me, and now I have a better idea of what's going on.

Several things come to mind:

* We should make sure this is _heavily_ tested, including the code that uses `ssh` and that talks over a Web socket.
* I think the code would be easier to read if every type and member didn't have the word "fuchsia" in it.
* Rather than `flutter_fuchsia`, the library name should either be more descriptive, e.g. `fuchsia_remote_debug_protocol`, or something more pithy that will become a name we use to refer to it unambiguously, like `fdb` (with a parallel to `adb`).
* We should find a solution that doesn't depend on the existence of a Fuchsia build tree. People will presumably develop for Fuchsia without having compiled it first.
* We should avoid depending on `logging` or any other package that isn't critical; our long-term plan is to dramatically scale back our dependencies because everything we depend on is something we pin for all our developers.
* We should support IPv6 from the get-go. Ideally we'd be agnostic as to v4 vs v6.
* `_potentiallyAvailablePort` is an antipattern; we've gone down that route several times in the past couple of years and it's always been a mistake. The race condition appears harmless but always becomes a flaky test or a crash or some such thing. We should allocate the port when we actually connect, not before then attempt to reuse it.
* I'm a little concerned about depending a magic file in `/tmp` on the host, but I presume that is something that will develop over time over in Fuchsia land.
* We should make sure this follows the [Flutter style guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo) and other conventions. For example, we should make sure to export everything (not just limited to `show`), have dartdocs for everything public (looks like this is already done), constructors first in classes, examples in .../examples/ rather than .../bin/, trailing commas in parameter declarations and argument lists so dartfmt does the right thing (basically an open `(`, `{`, or `[` should either be followed by a newline or the matching close should be on the same line; and vice versa with the closing equivalents).

MarkOSullivan94 commented:
@awdavies  / @Hixie could you provide a bit more clarity on what this is?

awdavies commented:
/CC @shrike69 @dglazkov

(more interested parties)

Hixie commented:
@MarkOSullivan94 As I understand it, this is a library for use by flutter_tools and flutter_driver (and eventually maybe standalone tools shipped with the library itself) for connecting to a Fuchsia host and listing the Flutter views instantiated on that host.

awdavies commented:
Sorry for the delay.  Addressing concerns.

> We should make sure this is heavily tested, including the code that uses ssh and that talks over a Web socket.

Working on that right now! :smile: 

> I think the code would be easier to read if every type and member didn't have the word "fuchsia" in it.

Definitely agree here, I'm just hoping to avoid stepping over existing namespaces (e.g. FlutterView, Isolate, VMService etc).

> We should find a solution that doesn't depend on the existence of a Fuchsia build tree. People will presumably develop for Fuchsia without having compiled it first.

Agreed.  My first instinct here was just to optionally point to an SSH config, though if there are more straightforward ways to connect to a Fuchsia instance I would definitely like to implement things that way instead.

> We should avoid depending on logging or any other package that isn't critical; our long-term plan is to dramatically scale back our dependencies because everything we depend on is something we pin for all our developers.

I can understand this, but is there a workaround so that we can support logging?  I think it is at least important to know if, say, the user doesn't have permission to bind to a port or the like.

> We should support IPv6 from the get-go. Ideally we'd be agnostic as to v4 vs v6.

Can do!

> `_potentiallyAvailablePort` is an antipattern; we've gone down that route several times in the past couple of years and it's always been a mistake. The race condition appears harmless but always becomes a flaky test or a crash or some such thing. We should allocate the port when we actually connect, not before then attempt to reuse it.

Ah.  This was something I blatantly stole from `flutter_tools`.  That race condition actually ended up being a huge problem on OSX for me, which is for the hack fallback code that I added.  I'll chop it out, though.

> I'm a little concerned about depending a magic file in /tmp on the host, but I presume that is something that will develop over time over in Fuchsia land.

Agreed.  The hope is to work with Fuchsia to turn this into an established service so that information about the VM instances can be taken from a websocket similar to the existing Dart VM.  The intent for now is that this method is a placeholder.

> We should make sure this follows the Flutter style guide and other conventions. For example, we should make sure to export everything (not just limited to show), have dartdocs for everything public (looks like this is already done), constructors first in classes, examples in .../examples/ rather than .../bin/, trailing commas in parameter declarations and argument lists so dartfmt does the right thing (basically an open (, {, or [ should either be followed by a newline or the matching close should be on the same line; and vice versa with the closing equivalents).

This was addressed recently, with the exceptions being ../bin/, exports, and trailing commas.  This along with testing and previous comments are in progress :smile:

awdavies commented:
> @awdavies / @Hixie could you provide a bit more clarity on what this is?

Long term this is intended to replace lots of code that _should_ be shared between flutter_driver and flutter_tools in such a way as to allow connecting to an arbitrary number of Dart VM's, and then on top of that allow for driving stories/mods on Fuchsia devices similar to flutter_driver.

I am unsure where the highest priorities lie regarding those requirements.  I am simply implementing things with guidance from @yjbanov.  I am also unsure of how much information has been sync'd between @yjbanov and @Hixie.  It appears they are both on the same page now?  I need to sync up with Yegor, though, to make sure I understand (as far as I understand there is no up-to-date PRD or design doc to reference).

At the time of writing this I am under the impression this is initially going to be more Fuchsia-centric code, but now I'm not sure given the existing feedback.

Hixie commented:
> I'm just hoping to avoid stepping over existing namespaces (e.g. FlutterView, Isolate, VMService etc).

Since this is in its own package I wouldn't worry too much about that.

> I can understand this, but is there a workaround so that we can support logging?  I think it is at least important to know if, say, the user doesn't have permission to bind to a port or the like.

Sure, just use `print` or provide a callback that you invoke. Logging doesn't have to be particularly complicated. :-)


> Ah.  This was something I blatantly stole from `flutter_tools`.

Oh dear, we still have code there that does this? I thought we'd excised all of it. Sigh. I'll file a bug or something to deal with that. In the meantime, yeah, let's avoid reproducing this mistake here.


Happy to give a more detailed review, let me know when it's ready. :-)

awdavies committed:
Refactors to remove redundant references.

awdavies committed:
Add in simple logging code.

Naming fixes.

awdavies committed:
Added missing file

awdavies committed:
SSH configs are now optional.

awdavies committed:
Remove port check race.

awdavies committed:
Adds an exception. Updates documentation.

awdavies committed:
More d11n updates

awdavies committed:
update pubspec.

awdavies committed:
Update packages. Address flutter_analyze issues.

awdavies committed:
Some more test boilerplate (doesn't work).

Also made ForwardedPort a public member (probably not necessary).

awdavies committed:
Tests SshCommandRunner. Puts in pipework for umbrella file test.

awdavies committed:
Refactoring for testing. Initial tests.

awdavies committed:
Adds test for dart_vm.  Linter addressed.

awdavies committed:
Adds dartvm timeout code and tests.

awdavies committed:
Addressed linter.

awdavies commented:
@Hixie @yjbanov PTAL

Supports IPv6 and has tests and such. Logging is self-contained, etc.  I've addressed all concerns AFAICT, though I'm unsure how well the code is tested regarding actual SSH.

yjbanov commented:
If @Hixie is ok with the high-level structure of this package, I can do a detailed code review.

Hixie commented:
General approach SGTM.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,126 @@
+/// logging function.
+class Logger {
+  /// Determines the tag included with the logging when chosen in the formatter
+  /// function.
```
I'm not sure I understand this sentence. I think all it wants to say is that all `[Log]` objects emitted by this logger will have `[Log.tag]` set to this value.

awdavies replied:
Fixed.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,126 @@
+  /// The human readable level of this log.
+  final String levelName;
+
+  /// The tag associated with the message.
```
We probably want to explain that this is the value of `[Logger.tag]` of the `Logger` that emitted this message.

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,126 @@
+  final String tag;
+
+  /// Determines what to do when the Logger creates and attempts to log a `Log`
+  /// object.
```
Use square brackets for references to classes.

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,126 @@
+  /// Determines what to do when the Logger creates and attempts to log a `Log`
+  /// object.
+  ///
+  /// The default format for this is to run the print function using the format
```
`[print]`

awdavies replied:
Fixed.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,126 @@
+/// Synchronously logs a `Log.`
+typedef void LoggingFunction(Log log);
+LoggingFunction _defaultFormatter = (Log log) {
+  print('[${log.levelName}] -- ${log.time}: ${log.message}');
```
Curious why we want to omit the `tag` by default.

awdavies replied:
Just figured it'd be nice to have optionally.  I added it in anyway

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,126 @@
+  ///
+  /// The default format for this is to run the print function using the format
+  /// string:
+  ///   '[${log.levelName}] -- ${log.time}: ${log.message}'
```
I think rather than documenting this here, we should make `_defaultFormatter` public and document everything there. This way the developer can reset the function back to the original implementation. However, I'd probably call it `defaultLoggingFunction` as "formatter" does not immediate convey that it may be used as a `LoggingFunction`.

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,126 @@
+  ///
+  /// This function can be reassigned to whatever functionality of your
+  /// choosing, so long as it has the same signature of `LoggingFunction` (it
+  /// can also be redefined as an async function, if doing file I/O, for
```
nit: let's spell out "asynchronous"
nit2: what does "redefined" mean? It sounds as if I can change the signature of the function.

Also, this seems to be in conflict with the dartdoc on the `LoggingFunction` typedef, which says that the function must be synchronous.

awdavies replied:
Fixed.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,126 @@
+  /// example).
+  static LoggingFunction loggingFunction = _defaultFormatter;
+
+  /// Determines the global logging level.
```
Instead of "global", I'd say that it "applies to all loggers".

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,126 @@
+  /// Determines the global logging level.
+  static LoggingLevel globalLevel = LoggingLevel.none;
+
+  /// Creates a logger with the given tag.
```
nit: ``[tag]``

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,126 @@
+  /// Creates a logger with the given tag.
+  Logger(this.tag);
+
+  /// Logs an severe level log.
```
nit: consider reducing the number of occurrences of the word "log" in the docs (can you can a can without a can? 😉). We should also provide a link to the corresponding `LoggingLevel` so that the reader could make a more informed choice which of the methods to use. E.g. "Logs a message under [LoggingLevel.severe]."

Also, nit: s/an/a/

(this comment applies to other methods too)

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,126 @@
+  /// Logs no logs.
+  none,
+
+  /// Logs severe logs at the most.
```
I think these docs should provide some guidance on how these levels should be treated by, e.g. a custom implementation of `LoggingFunction`. Typically "severe" means that an unrecoverable error has been encountered and the process will or should quit. "warning" means that an error happened but the process will attempt to continue and may or may not succeed. The default implementation of `LoggingFunction` should also uphold this contract, (e.g. quit or throw upon "severe" error).

awdavies replied:
Done.  Also implemented `defaultLoggingFunction` to exit in the event of a severe log, and changed the logic so that severe logs are always logged.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,126 @@
+  void severe(String message) {
+    if (globalLevel.index >= LoggingLevel.severe.index) {
+      loggingFunction(
+          new Log(message, tag, new DateTime.now(), LoggingLevel.severe));
```
Why not populate the time in the constructor of `Log`?

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,126 @@
+/// as an enum, the name of the level as a string, and a tag. This class is used
+/// to print from the global logging function defined in
+/// `Logger.loggingFunction` (a function that can be user-defined).
+class Log {
```
nit: to me a "log" contains many messages. I'd call this "LogRecord" or "LogMessage".

awdavies replied:
Implemented as LogMessage

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/common/network.dart:
```diff
@@ -0,0 +1,34 @@
+
+import 'dart:core';
+
+/// Determines whether the address is valid IPv6 or IPv4 format.
```
nit: s/is/has/

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/common/network.dart:
```diff
@@ -0,0 +1,34 @@
+  }
+}
+
+/// Returns true if the address is a valid IPv6 address.
```
nit: for consistency, we should probably stick with one of "has format" or "is address".

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/dart/dart_vm.dart:
```diff
@@ -0,0 +1,180 @@
+
+final Logger _log = new Logger('DartVm');
+
+/// Definition of an asynchronous function for connecting to a `Uri`.
```
nit: s/Definition/Signature/

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/dart/dart_vm.dart:
```diff
@@ -0,0 +1,180 @@
+
+/// Definition of an asynchronous function for connecting to a `Uri`.
+///
+/// Creates a JSON RPC-2 connection.
```
I'd roll this into the paragraph above by replacing "connecting" with "establishing a JSON RPC-2 connection".

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/dart/dart_vm.dart:
```diff
@@ -0,0 +1,180 @@
+/// Creates a JSON RPC-2 connection.
+typedef Future<json_rpc.Peer> RpcPeerConnectionFunction(Uri uri);
+
+/// `DartVm` uses this function to connect to the Dart VM on Fuchsia.
```
use square brackets to refer to classes - `[DartVm]`.

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/dart/dart_vm.dart:
```diff
@@ -0,0 +1,180 @@
+
+/// `DartVm` uses this function to connect to the Dart VM on Fuchsia.
+///
+/// This function can be assigned to a different one out in the event that a
```
Remove "out"?

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/dart/dart_vm.dart:
```diff
@@ -0,0 +1,180 @@
+  return attemptConnection(uri);
+}
+
+/// Restores the VM service connection function to the original state.
```
nit: s/original state/default implementation/ (it's not clear what "state" means for a function).

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/dart/dart_vm.dart:
```diff
@@ -0,0 +1,180 @@
+}
+
+/// Simple JSON RPC peer with a Fuchsia Dart VM service instance. Wraps most
+/// methods around a JSON RPC peer class.
```
nit: https://www.dartlang.org/guides/language/effective-dart/documentation#do-make-the-first-sentence-a-standalone-paragraph

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/dart/dart_vm.dart:
```diff
@@ -0,0 +1,180 @@
+
+  /// Invokes a raw JSON RPC command with the VM service.
+  ///
+  /// When `timeout` is set and reached, throws a `TimeoutException`.
```
ditto: backticks for parameter, square brackets for everything else (classes, fields, methods, functions).

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/dart/dart_vm.dart:
```diff
@@ -0,0 +1,180 @@
+    return views;
+  }
+
+  /// Shuts down all active connections.
```
"Disconnects from the Dart VM service"? :) 

We should probably mention that this object is no longer usable after calling this method.

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/dart/dart_vm.dart:
```diff
@@ -0,0 +1,180 @@
+
+/// Represents an instance of a Flutter view running on a Fuchsia device.
+class FlutterView {
+  /// Determines the name of the Isolate associated with this view. If there is
```
nit: "isolate" (it's not a proper name)

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/dart/dart_vm.dart:
```diff
@@ -0,0 +1,180 @@
+/// Represents an instance of a Flutter view running on a Fuchsia device.
+class FlutterView {
+  /// Determines the name of the Isolate associated with this view. If there is
+  /// no associated Isolate, this will be set to the view's ID.
```
ditto

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/dart/dart_vm.dart:
```diff
@@ -0,0 +1,180 @@
+  /// null.
+  ///
+  /// All other cases return a `FlutterView` instance. The name of the
+  /// view may be null, but the id will always be set.
```
nit: "ID"

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/dart/dart_vm.dart:
```diff
@@ -0,0 +1,180 @@
+  /// The ID of the `FlutterView`.
+  String get id => _id;
+
+  /// Returns the name of the `FucshiaFlutterView`.
```
There is no more FucshiaFlutterView right?

awdavies replied:
Woops! Looks like sed missed something.

yjbanov reviewed (commented):
Reviewed about 70%. Will continue tomorrow.

awdavies commented on packages/fuchsia_remote_debug_protocol/examples/list_vms_and_flutter_views.dart:
```diff
@@ -0,0 +1,36 @@
+import 'dart:core';
+
+import '../lib/fuchsia_remote_debug_protocol.dart';
+import '../lib/lib_logging.dart' as lib_logging;
```
Not in the example code, no

awdavies replied:
Fixed.

awdavies commented on packages/fuchsia_remote_debug_protocol/lib/lib_logging.dart:
```diff
@@ -0,0 +1,12 @@
+// Copyright 2018 The Flutter Authors. All rights reserved.
```
Would it be alright just to name it logging?  The intent wasn't to prefix it with "lib_" but to say that the file is a library logging utility so that the user can set a logging function for the library/package.

awdavies replied:
Renamed to logging.  I figure the documentation and the fact that it is part of the package should imply that it is for internal logging of the package itself.

awdavies commented on packages/fuchsia_remote_debug_protocol/examples/list_vms_and_flutter_views.dart:
```diff
@@ -0,0 +1,36 @@
+/// Runs through a simple usage of the fuchsia_remote_debug_protocol library:
+/// connects to a remote machine at the address in argument 1 (interface
+/// optional for argument 2) to list all active flutter views and Dart VM's
+/// running on said device.
```
Done.

awdavies commented on packages/fuchsia_remote_debug_protocol/examples/list_vms_and_flutter_views.dart:
```diff
@@ -0,0 +1,36 @@
+import 'dart:core';
+
+import '../lib/fuchsia_remote_debug_protocol.dart';
+import '../lib/lib_logging.dart' as lib_logging;
```
Done.

awdavies commented on packages/fuchsia_remote_debug_protocol/examples/list_vms_and_flutter_views.dart:
```diff
@@ -0,0 +1,36 @@
+  lib_logging.Logger.globalLevel = lib_logging.LoggingLevel.info;
+  final String address = args[0];
+  final String interface = args.length > 1 ? args[1] : '';
+  // Example ssh config path for the fuchsia device after having made a local
```
Done.

awdavies committed:
Addressed review comments.

awdavies commented on packages/fuchsia_remote_debug_protocol/lib/lib_logging.dart:
```diff
@@ -0,0 +1,12 @@
+/// the event that you are using a logging package with similar class names.
+library lib_logging;
+
+export 'src/common/logging.dart';
```
This is just based off of one of @Hixie 's comments:
>We should make sure this follows the Flutter style guide and other conventions. For example, we should make sure to export everything (not just limited to show)

awdavies commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,126 @@
+  all,
+}
+
+/// Synchronously logs a `Log.`
```
Fixed.

awdavies commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,126 @@
+/// Includes a message, the time the message was created, the level of the log
+/// as an enum, the name of the level as a string, and a tag. This class is used
+/// to print from the global logging function defined in
+/// `Logger.loggingFunction` (a function that can be user-defined).
```
Done.

awdavies commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,126 @@
+  print('[${log.levelName}] -- ${log.time}: ${log.message}');
+};
+
+/// Represents a Log created by the logger.
```
Done.

awdavies commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,126 @@
+/// logging level.
+///
+/// Example of setting log level to `LoggingLevel.warning` and creating a
+/// logging function.
```
Fixed.

awdavies commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,126 @@
+            level.toString().substring(level.toString().indexOf('.') + 1);
+}
+
+/// Very barebones logging class. Prints using the global LoggingFunction and
```
Fixed.

awdavies committed:
Remaining comments addressed. Documentation fixed.

awdavies committed:
Updated packages again.

awdavies committed:
Updated packages one more time.

awdavies committed:
Addresses linter issues.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/runners/ssh_command_runner.dart:
```diff
@@ -0,0 +1,99 @@
+
+  @override
+  String toString() {
+    return 'SshCommandError: $message\n${super.stackTrace}';
```
hint: You can $ in front of SshCommandError, which will produce the same string but protect from spelling mistakes or future class renames.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/runners/ssh_command_runner.dart:
```diff
@@ -0,0 +1,99 @@
+  }
+}
+
+/// Runs a command remotely on a Fuchsia device. Requires a fuchsia root and
```
Did you mean "commands"?

Also, nit: make the first sentence a standalone paragraph.

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/runners/ssh_command_runner.dart:
```diff
@@ -0,0 +1,99 @@
+}
+
+/// Runs a command remotely on a Fuchsia device. Requires a fuchsia root and
+/// build type (to load the ssh config), and the address of the fuchsia
```
nit: fuchsia -> Fuchsia (in two places)

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/runners/ssh_command_runner.dart:
```diff
@@ -0,0 +1,99 @@
+      args.add(address);
+    }
+    args.add(command);
+    _log.fine(args.join(' '));
```
This log message could be more descriptive too.

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/test/fuchsia_remote_connection_test.dart:
```diff
@@ -0,0 +1,150 @@
+    });
+
+    tearDown(() {
+      /// Most functions will mock out the port forwarding and connection
```
nit: most tests?

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/test/src/common/matchers.dart:
```diff
@@ -0,0 +1,9 @@
+import 'package:test/test.dart';
+
+/// Wrapper for catching an [ArgumentError] being thrown.
+final Matcher throwsArgumentError =
```
`package:test` already has this: https://www.dartdocs.org/documentation/test/0.12.31/test/throwsArgumentError-constant.html

awdavies replied:
Must have not worked from a typo before.  Removed.

yjbanov commented on packages/fuchsia_remote_debug_protocol/lib/src/dart/dart_vm.dart:
```diff
@@ -0,0 +1,183 @@
+  /// Attempts to construct a [FlutterView] from a json representation.
+  ///
+  /// If there is no isolate and no ID for the view, returns null. If there is
+  /// an associated isolate, and there is name for said isolate, also returns
```
Did you mean "but there is no name"?

awdavies replied:
Yep.  Fixed.

yjbanov commented on packages/fuchsia_remote_debug_protocol/test/src/dart/dart_vm_test.dart:
```diff
@@ -0,0 +1,181 @@
+      expect(views[2].name, 'file://flutterBinary2');
+    });
+
+    test('invalid flutter view parsing', () async {
```
Are they invalid or just happen to not be Flutter views? If they are actually invalid, perhaps we should throw an error (after all this will be used for testing Fuchsia).

awdavies replied:
I don't think those are mutually exclusive given the RPC invocation is intended to get only Flutter views.

awdavies replied:
I've updated the code to throw an `RpcFormatError` when receiving a malformed response.

yjbanov commented on packages/fuchsia_remote_debug_protocol/test/src/runners/ssh_command_runner_test.dart:
```diff
@@ -0,0 +1,142 @@
+      await runner.run('ls /whatever');
+      final List<String> passedCommand =
+          verify(mockProcessManager.run(captureAny)).captured.single;
+      expect(passedCommand.indexOf('$ipV6Addr%$interface'), isNot(-1));
```
See also https://www.dartdocs.org/documentation/test/0.12.31/test/contains.html (this comment applies throughout the file)

awdavies replied:
Done.

yjbanov commented on packages/fuchsia_remote_debug_protocol/test/src/runners/ssh_command_runner_test.dart:
```diff
@@ -0,0 +1,142 @@
+          lines''');
+      when(mockProcessResult.exitCode).thenReturn(0);
+      final List<String> result = await runner.run('oihaw');
+      expect(result.length, 4);
```
See also https://www.dartdocs.org/documentation/test/0.12.31/test/hasLength.html (when it fails, it produces a better error message)

awdavies replied:
Done.

awdavies commented on packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart:
```diff
@@ -0,0 +1,340 @@
+  ///
+  /// Accepts an `address` to a Fuchsia device, and requires a root
+  /// directory in which the Fuchsia Device was built (along with the
+  /// `buildType`) in order to open the associated ssh_config for port
```
I think this comment is on an outdated commit.

awdavies commented on packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart:
```diff
@@ -0,0 +1,340 @@
+  /// Accepts an `address` to a Fuchsia device, and requires a root
+  /// directory in which the Fuchsia Device was built (along with the
+  /// `buildType`) in order to open the associated ssh_config for port
+  /// forwarding. Will throw an `ArgumentError` if the address is malformed.
```
Same as above.

awdavies commented on packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart:
```diff
@@ -0,0 +1,336 @@
+  ///
+  /// Accepts an `address` to a Fuchsia device, and optionally a `sshConfigPath` in order to open the
+  /// associated ssh_config for port forwarding.
+  /// Will throw an [ArgumentError] if `address` is malformed.
```
Done.

awdavies commented on packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart:
```diff
@@ -0,0 +1,336 @@
+      final DartVm vmService = await _getDartVm(fp.port);
+      views.addAll(await vmService.getAllFlutterViews());
+    }
+    return views;
```
Done.

awdavies commented on packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart:
```diff
@@ -0,0 +1,336 @@
+    return _dartVmCache[port];
+  }
+
+  /// Forwards a series of local device ports to the.
```
Done.

awdavies commented on packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart:
```diff
@@ -0,0 +1,336 @@
+/// which a connection is persisted along the lifetime of this object.
+///
+/// When a port forwarder is shut down it must use its `stop` function to clean
+/// up.
```
Done.

awdavies commented on packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart:
```diff
@@ -0,0 +1,336 @@
+/// Instances of this class represent a running ssh tunnel.
+///
+/// The SSH tunnel is from the host to a VM service running on a Fuchsia device.
+class _SshPortForwarder extends PortForwarder {
```
Done.

awdavies commented on packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart:
```diff
@@ -0,0 +1,336 @@
+  Future<Null> stop();
+}
+
+/// Instances of this class represent a running ssh tunnel.
```
Done.

awdavies commented on packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart:
```diff
@@ -0,0 +1,336 @@
+  }
+
+  /// Kills the SSH forwarding command, then to ensure no ports are forwarded,
+  /// runs the ssh 'cancel' command to shut down port forwarding completely.
```
Done

awdavies commented on packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart:
```diff
@@ -0,0 +1,336 @@
+  @override
+  Future<Null> stop() async {
+    // Kill the original ssh process if it is still around.
+    _process?.kill();
```
The original code this was copied from included a null check here. Fixed.

awdavies commented on packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart:
```diff
@@ -0,0 +1,336 @@
+      targetAddress,
+    ]);
+    final ProcessResult result = await _processManager.run(command);
+    _log.fine(command.join(' '));
```
Done.

awdavies commented on packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart:
```diff
@@ -0,0 +1,336 @@
+    }
+  }
+
+  static Future<int> _potentiallyAvailablePort() async {
```
Done.

awdavies commented on packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart:
```diff
@@ -0,0 +1,336 @@
+      // Failures are signaled by a return value of 0 from this function.
+      _log.warning('_potentiallyAvailablePort failed: $e');
+    }
+    await s?.close();
```
Done.

awdavies commented on packages/fuchsia_remote_debug_protocol/lib/src/runners/ssh_command_runner.dart:
```diff
@@ -0,0 +1,99 @@
+
+/// An error raised when a command fails to run within the [SshCommandRunner].
+///
+/// Note that this occurs for both connection failures, and for failure to
```
Done.

awdavies commented on packages/fuchsia_remote_debug_protocol/lib/src/runners/ssh_command_runner.dart:
```diff
@@ -0,0 +1,99 @@
+/// An error raised when a command fails to run within the [SshCommandRunner].
+///
+/// Note that this occurs for both connection failures, and for failure to
+/// running the command on the remote device. This error is raised when the
```
Done.

awdavies committed:
Addressed more code review comments.

awdavies committed:
Added RpcFormatError for malformed RPC responses w/ Dart VM

awdavies commented:
@yjbanov Comments addressed.  PTAL

yjbanov commented on packages/fuchsia_remote_debug_protocol/examples/list_vms_and_flutter_views.dart:
```diff
@@ -0,0 +1,44 @@
+// Copyright 2018 The Flutter Authors. All rights reserved.
```
Actually, we already have an `examples` directory at the root of the repository meant to demo all kinds of stuff. This should probably go there.

awdavies replied:
That directory contains demos of flutter itself (it even says that you should run `flutter run` inside example directories in the readme), so I don't think that would be appropriate.

yjbanov replied:
Ah, got it.

yjbanov reviewed (approved):
This LGTM if LGT @Hixie.

Before submitting please make sure:

- [ ] the unit tests run as part of our Travis script
- [ ] the package is captured by the analyzer
- [ ] all types that are returned by public API are also exported through libraries under `lib/` (so that a user does not have to import from `lib/src`)

awdavies committed:
Addressed linter.  Added travis test invocation.  Updated pkgs.

awdavies commented:
-[x] the unit tests run as part of our Travis script
-[x] the package is captured by the analyzer
-[x]  all types that are returned by public API are also exported through libraries under `lib/` (so that a user does not have to import from `lib/src`)

awdavies committed:
Fixed internal analyzer issues.

awdavies commented:
@yjbanov @Hixie there are some minor analyzer issues (but these are due to the flutter package.  `slider_theme.dart` is also missing documented public members).

awdavies committed:
updated example to reflect comments.

yjbanov commented:
@Hixie, did you want to do a code style pass on this PR, or is it good to merge?

Hixie commented on packages/fuchsia_remote_debug_protocol/examples/list_vms_and_flutter_views.dart:
```diff
@@ -0,0 +1,47 @@
+/// Runs through a simple usage of the fuchsia_remote_debug_protocol library:
+/// connects to a remote machine at the address in argument 1 (interface
+/// optional for argument 2) to list all active flutter views and Dart VM's
+/// running on said device. Note that this uses an SSH config file (optional,
```
nit: strike "note that", it doesn't add anything.

awdavies replied:
Fixed.

Hixie commented on packages/fuchsia_remote_debug_protocol/examples/list_vms_and_flutter_views.dart:
```diff
@@ -0,0 +1,47 @@
+
+/// Runs through a simple usage of the fuchsia_remote_debug_protocol library:
+/// connects to a remote machine at the address in argument 1 (interface
+/// optional for argument 2) to list all active flutter views and Dart VM's
```
VM's -> VMs

awdavies replied:
Fixed.

Hixie commented on packages/fuchsia_remote_debug_protocol/examples/list_vms_and_flutter_views.dart:
```diff
@@ -0,0 +1,47 @@
+  // information.
+  Logger.globalLevel = LoggingLevel.info;
+  if (args.length == 0) {
+    return;
```
maybe print a message before exiting?

awdavies replied:
Done.

Hixie commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,149 @@
+/// Determines the level of logging.
+///
+/// Verbosity is increasing from zero (none) to four (fine). The fifth level
+/// (all) logs everything.
```
`all` is the sixth level.

Hixie commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,149 @@
+  /// Logs no logs.
+  none,
+
+  /// Logs severe messages at the most (note that severe messages are always
```
strike "note that"

awdavies replied:
Done.

Hixie commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,149 @@
+  ///
+  /// Severe means that the process has
+  /// encountered a critical level of failure in which it cannot recover and
+  /// will terminate as a result.
```
rewrap

awdavies replied:
Done.

Hixie commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,149 @@
+/// Example of setting log level to [LoggingLevel.warning] and creating a
+/// logging function:
+///
+/// Logger.globalLevel = LoggingLevel.warning;
```
wrap examples in triple backtick blocks, tagged as "dart", as in:
```
'''dart
// example here
'''
```
...but using backticks instead of apostrophes

awdavies replied:
Done.

Hixie commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,149 @@
+  static LoggingLevel globalLevel = LoggingLevel.none;
+
+  /// Creates a logger with the given [tag].
+  Logger(this.tag);
```
constructors should be first, so that we can quickly tell if there's an implicit no-argument constructor without reading the whole definition of the class

awdavies replied:
Fixed.

Hixie commented on packages/fuchsia_remote_debug_protocol/lib/src/common/logging.dart:
```diff
@@ -0,0 +1,149 @@
+
+  /// Logs a [LoggingLevel.severe] level `message`.
+  ///
+  /// Note that severe messages are always logged.
```
strike "note that"

awdavies replied:
Done.

Hixie commented on packages/fuchsia_remote_debug_protocol/lib/src/dart/dart_vm.dart:
```diff
@@ -0,0 +1,196 @@
+  final String message;
+
+  /// Basic constructor outlining the reason for the format error.
+  RpcFormatError(this.message);
```
constructor first

awdavies replied:
Fixed.

Hixie commented on packages/fuchsia_remote_debug_protocol/lib/src/dart/dart_vm.dart:
```diff
@@ -0,0 +1,196 @@
+class DartVm {
+  final json_rpc.Peer _peer;
+
+  DartVm._(this._peer);
```
constructor first

awdavies replied:
Fixed.

Hixie commented on packages/fuchsia_remote_debug_protocol/lib/src/dart/dart_vm.dart:
```diff
@@ -0,0 +1,196 @@
+  ///
+  /// All other cases return a [FlutterView] instance. The name of the
+  /// view may be null, but the id will always be set.
+  factory FlutterView._fromJson(Map<String, dynamic> json) {
```
constructor first

awdavies replied:
Fixed.

Hixie commented on packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart:
```diff
@@ -0,0 +1,342 @@
+  /// Finds the port by binding a socket to zero on the loopback interface, and
+  /// then unbinding. If this action could not be run for whatever reason, will
+  /// log a warning and return zero.
+  static Future<int> _potentiallyAvailablePort() async {
```
this is a race condition. we should allocate a port once, not twice hoping that no other process allocates it in the meantime.

awdavies replied:
So after snooping around, while it is a race condition, it seems there's no way to locally forward a port through SSH without previous knowledge of said port.

Would it be reasonable to implement this without use of SSH?  I believe it should be possible, just without secure data transfer.

awdavies replied:
Fixed with bound socket.

Hixie commented on packages/fuchsia_remote_debug_protocol/lib/src/runners/ssh_command_runner.dart:
```diff
@@ -0,0 +1,100 @@
+
+  /// Basic constructor outlining the reason for the SSH command failure through
+  /// the message string.
+  SshCommandError(this.message);
```
constructor first

Hixie commented on packages/fuchsia_remote_debug_protocol/lib/src/runners/ssh_command_runner.dart:
```diff
@@ -0,0 +1,100 @@
+  /// IPv4 nor IPv6. Note that when connecting to a link local address (fe80::
+  /// is usually at the start of the address), then an interface should be
+  /// supplied.
+  SshCommandRunner({this.address, this.interface = '', this.sshConfigPath})
```
constructors first

Hixie commented on packages/fuchsia_remote_debug_protocol/pubspec.yaml:
```diff
@@ -0,0 +1,66 @@
+  mockito: 2.2.3
+  test: 0.12.32+1
+
+  args: 1.4.1 # TRANSITIVE DEPENDENCY
```
please re-run the script, it was modified recently so that some of these will now end up in the "dependencies" section rather than "dev_dependencies"

awdavies replied:
Done.

Hixie commented on packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart:
```diff
@@ -0,0 +1,342 @@
+          '$address:$remotePort');
+      return null;
+    }
+    // TODO: The square-bracket enclosure for using the IPv6 loopback didn't
```
syntax for TODOs is "// TODO(username): description"

Hixie commented on packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart:
```diff
@@ -0,0 +1,342 @@
+  final String _interface;
+  final bool _ipV6;
+
+  _SshPortForwarder._(
```
constructor first

awdavies replied:
Fixed.

Hixie commented on packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart:
```diff
@@ -0,0 +1,342 @@
+  /// attempting to acquire the ports.
+  Future<List<int>> getDeviceServicePorts() async {
+    final List<String> lsOutput =
+        await _sshCommandRunner.run('ls /tmp/dart.services');
```
my understanding is that this feature's days are numbered. maybe add a TODO saying we'll soon have to fix this.

awdavies replied:
Correct.  The plan was to figure out/change this shortly after submitting this change.  TODO added.

Hixie commented on packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart:
```diff
@@ -0,0 +1,342 @@
+        new SshCommandRunner(
+            address: address,
+            interface: interface,
+            sshConfigPath: sshConfigPath));
```
if you use dartfmt, add trailing commas before each `)` and `]` and rewrap this.

It should look like this to match Flutter style:

```dart
static Future<FuchsiaRemoteConnection> connect(
  String address, [
  String interface = '',
  String sshConfigPath,
]) async {
  return await FuchsiaRemoteConnection.connectWithSshCommandRunner(
    new SshCommandRunner(
      address: address,
      interface: interface,
      sshConfigPath: sshConfigPath,
    ),
  );
}
```

awdavies replied:
Fixed.

Hixie commented on packages/fuchsia_remote_debug_protocol/lib/src/fuchsia_remote_connection.dart:
```diff
@@ -0,0 +1,342 @@
+  /// calls. Keys a forwarded port to a DartVm connection instance.
+  final Map<int, DartVm> _dartVmCache = <int, DartVm>{};
+
+  FuchsiaRemoteConnection._(this._useIpV6Loopback, this._sshCommandRunner);
```
constructors first

awdavies replied:
Fixed.

Hixie commented on packages/fuchsia_remote_debug_protocol/lib/src/dart/dart_vm.dart:
```diff
@@ -0,0 +1,196 @@
+    return new FlutterView._(name, id);
+  }
+
+  FlutterView._(this._name, this._id);
```
constructors first

awdavies replied:
Fixed.

Hixie commented on packages/fuchsia_remote_debug_protocol/lib/src/dart/dart_vm.dart:
```diff
@@ -0,0 +1,196 @@
+  ///
+  /// When `timeout` is set and reached, throws a [TimeoutException].
+  ///
+  /// If the function returns, it is with a raw JSON response.
```
a raw JSON response would be a string, this returns a map. Maybe "a parsed JSON object"?

awdavies replied:
Fixed.

Hixie commented on packages/fuchsia_remote_debug_protocol/lib/fuchsia_remote_debug_protocol.dart:
```diff
@@ -0,0 +1,15 @@
+/// and emulators.
+///
+/// The application typically runs in a separate process from the actual
+/// test, wherein the user can supply the process with various events to test
```
if it's on a remote device, isn't it guaranteed to be in a separate process?

awdavies replied:
I don't remember the particulars, but I remember being told it won't always be on a remote device.

Hixie commented on packages/fuchsia_remote_debug_protocol/lib/fuchsia_remote_debug_protocol.dart:
```diff
@@ -0,0 +1,15 @@
+/// The application typically runs in a separate process from the actual
+/// test, wherein the user can supply the process with various events to test
+/// the behavior of said application.
+library fuchsia_remote_debug_protocol;
```
we should include more details here, such as a high-level overview of the general shape of the API, where to start when using this API, and so forth.

awdavies replied:
Added.

Hixie commented on packages/fuchsia_remote_debug_protocol/lib/fuchsia_remote_debug_protocol.dart:
```diff
@@ -0,0 +1,15 @@
+
+export 'src/dart/dart_vm.dart';
+export 'src/fuchsia_remote_connection.dart';
+export 'src/runners/ssh_command_runner.dart';
```
this should export every file in the package, ideally.

Hixie replied:
having logging exported separately is reasonable, but e.g. the network file should be exported

awdavies replied:
Fixed.

Hixie commented:
Just some style nits and one race condition that we should fix before landing (we've had huge troubles with race conditions like this causing flakes, and I really don't want to go down that road again...).

awdavies commented on packages/fuchsia_remote_debug_protocol/lib/src/dart/dart_vm.dart:
```diff
@@ -0,0 +1,196 @@
+      throw new TimeoutException(
+          'Peer connection timed out during RPC call', timeout);
+    });
+  }
```
Fixed.

awdavies committed:
Addressed review comments regarding style.

awdavies committed:
Ran package upgrade/rebased.

awdavies committed:
Update forwarding to avoid race condition.

awdavies committed:
Null check and linter fixes.

awdavies committed:
Rebase and pub upgrade.

awdavies committed:
minor formatting. test fixes.

awdavies committed:
More analyzer fixes.

awdavies commented:
@Hixie @yjbanov 

Everything should be ready to go!

awdavies commented:
Oh wait... had to fix a constructor.  Should be good.

awdavies committed:
package update

Hixie commented:
![LGTM](http://junkyard.damowmow.com/577)

awdavies committed:
Fixed pubspec for package.

awdavies committed:
Constructor-first fix.

shrike69 commented:
Great!
